exercises/xorcism/tests/xorcism.rs

```diff
@@ -0,0 +1,303 @@
+use rstest::rstest;
+use rstest_reuse::{self, *};
+#[cfg(feature = "io")]
+use std::io::{Read, Write};
+use xorcism::Xorcism;
+use hexlit::hex;
+
+#[test]
+fn identity() {
+    let mut xs = Xorcism::new(&[0]);
+    let data = "This is super-secret, cutting edge encryption, folks.";
+
+    assert_eq!(
+        xs.munge(data.as_bytes()).collect::<Vec<_>>(),
+        data.as_bytes()
+    );
+}
+
+#[test]
+#[ignore]
+fn munge_output_has_len() {
+    let mut xs = Xorcism::new(&[0]);
+    let data = "The output must have a definite length";
+    assert_eq!(xs.munge(data.as_bytes()).len(), 38);
+}
+
+#[test]
+#[ignore]
+fn statefulness() {
+    // we expect Xorcism to be stateful: at the end of a munging run, the key has rotated.
+    // this means that until the key has completely rotated around, equal inputs will produce
+    // unequal outputs.
+    let key = &[0, 1, 2, 3, 4, 5, 6, 7];
+    let input = &[0b1010_1010, 0b0101_0101];
+
+    let mut xs = Xorcism::new(&key);
+    let out1: Vec<_> = xs.munge(input).collect();
+    let out2: Vec<_> = xs.munge(input).collect();
```

Comment:
A student who reads only up to this point in the code and hasn't read the later parts may have a question about an ambiguity of multiple calls to munge. They might ask this question:

So, let's say you have two separate in-flight munges:

```rust
let mut iter1 = xs.munge(input);
let mut iter2 = xs.munge(input);
println!("1 has {}", iter1.next().unwrap());
println!("2 has {}", iter2.next().unwrap());
println!("1 has {}", iter1.next().unwrap());
println!("2 has {}", iter2.next().unwrap());
```

Given that the calls to `next()` were interleaved between the two, what should have happened?

1. `iter1` XORs its input with 0 and 1 and `iter2` XORs its input with 2 and 3, just as they would have if the calls were not itnerleaved
2. `iter1` XORs its input with 0 and 2 and `iter2` XORs its input with 1 and 3, reflecting the fact that they were interleaved

Now, behaviour 2 would be extremely difficult to implement (it implies that `out1` and `out2` are mutating some state common to both of them). And, in fact, we test behaviour 1 later on in the round-tripping tests, since we are chaining two iterator adaptors together and the round tripping will only work correctly with behaviour 1.

The only possible concern here, therefore, is if a student does all three of the following:

* reads this test
* doesn't read the round-tripping tests
* somehow misinterprets the test to understand that behaviour 2 is expected.

I cannot predict how likely this scenario will be, but it requires three conditions simultaneously so it could be pretty rare. So I would be happy if the answer here is "take no action for now". However I can also suggest doing one of the following two:

1. Make some sort of comment here making clear that this test isn't testing for behaviour 2. Not sure exactly how to word such a comment, though.
2. Before this test, place a more explicit test of interleaving that makes clear what the right behaviour is

Reply:
EBWODP:

> And, in fact, we test behaviour 1 later on in the round-tripping tests, since we are chaining two iterator adaptors together and the round tripping will only work correctly with behaviour 1.

What I said didn't make sense; it would only be testing behaviour1 if the key length evenly divides the text length. In fact the round-tripping uses two instances of Xorcism.

So we are not testing either behaviour 1 or 2 right now. So we could explicitly say that such behaviour is not being tested at all, explicitly test the behaviour that is actually plausible to implement, or do nothing.

Reply:
You can't have two separate in-flight munges: `munge` takes `&mut self`.

Reply:
Ooh. That settles things quite nicely, then.

Now, I'm making a guess here that we know this because we know what type `munge` must eventually return (I don't know for sure yet since I haven't completed my implementation, but I know what lifetime I need it to have).

If we start from just the stub signature given to the student:

```rust
pub fn munge<Data>(&mut self, data: Data) -> impl MungeOutput
```

They may not yet know at that point that you cannot have two in-flight munges, and (as I just checked), such a test compiles with the stub signature (though presumably not with the final signature).

So, mentioning that it's not going to be tested could still have value. But leaving things as-is could also be a useful learning experience. I don't feel strongly about it, so perhaps we leave it as-is unless finding out from student feedback that a change is needed.

Reply:
My inclination in this case is to leave things as-is, changing it if we receive student reports that they're confused on that point.